### PR TITLE
Create probe0.go

### PR DESCRIPTION
**Pull request by alicjakwie**
CLOSES (#47 )

### CODE REVIEW

File: GCS/probe0/probe0.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+// Probe holds aggregate information about all probe runs, per-target.
+type Probe struct {
+	name    string
+	c       *ProbeConf
+	targets []string
+	opts    *options.Options
+
+	res map[string]*metrics.EventMetrics // Results by target
```
Comment (deniadelcid) [inline GCS/probe0/probe0.go]:
```suggestion
	res map[string]*metrics.EventMetrics  // Results by target
```

File: GCS/probe0/probe0.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	"cloud.google.com/go/storage"
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/google/cloudprober/targets/endpoint"
+	"github.com/googleinterns/step224-2020/blob/evan/hermesMain/cmd"
+)
+
+// Probe holds aggregate information about all probe runs, per-target.
+type Probe struct {
```
Comment (deniadelcid) [inline GCS/probe0/probe0.go]: Probe sounds too generic, consider renaming it to `HermesProbe`

File: GCS/probe0/probe0.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+	res map[string]*metrics.EventMetrics // Results by target
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
+func (p *Probe) Init(name string, opts *options.Options) error {
+	c, ok := opts.ProbeConf.(*ProbeConf)
+	if !ok {
+		return fmt.Errorf("not a my probe config")
```
Comment (deniadelcid) [inline GCS/probe0/probe0.go]: Maybe rephrase this error to: "Invalid Hermes probe configuration"

File: GCS/probe0/probe0.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	targets []string
+	opts    *options.Options
+
+	res map[string]*metrics.EventMetrics // Results by target
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
+func (p *Probe) Init(name string, opts *options.Options) error {
+	c, ok := opts.ProbeConf.(*ProbeConf)
```
Comment (deniadelcid) [inline GCS/probe0/probe0.go]: The syntax for `opt.ProbeConf.(*ProbeConf)` seems odd. Does it run?

File: GCS/probe0/probe0.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	if !ok {
+		return fmt.Errorf("not a my probe config")
+	}
+	p.c = c
```
Comment (deniadelcid) [inline GCS/probe0/probe0.go]:
```suggestion
```

File: GCS/probe0/probe0.go (hunk 4 above)
Comment (deniadelcid) [inline GCS/probe0/probe0.go]:
```suggestion
	p.c, ok := opts.ProbeConf.(*ProbeConf)
```

File: GCS/probe0/probe0.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			// On probe tick, write data to the channel and run probe.
+			for _, em := range p.res {
+				dataChan <- em
+			}
+			p.targets = endpoint.NamesFromEndpoints(p.opts.Targets.ListEndpoints())
+			p.initProbeMetrics()
+			probeCtx, cancelFunc := context.WithDeadline(ctx, time.Now().Add(p.opts.Timeout))
```
Comment (deniadelcid) [inline GCS/probe0/probe0.go]: Consider making cancelFunc a struct field for Probe, this would allow us to actually cancel the probe from outside

**Review by deniadelcid — COMMENTED** [review]
- [ ] Run gofmt for this file.

- [ ]  Add a package to the file.

- [ ]  Can you add a file with your unit tests for probe0.go?

File: GCS/probe0/probe0.go (hunk 2 above)
Comment (alicjakwie) [reply GCS/probe0/probe0.go]:
```suggestion
type HermesProbe struct {
```

File: GCS/probe0/probe0.go (hunk 3 above)
Comment (alicjakwie) [reply GCS/probe0/probe0.go]:
```suggestion
		return fmt.Errorf("Invalid Hermes probe configuration")
```

File: GCS/probe0/probe0.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	c       *ProbeConf
+	targets []string
+	opts    *options.Options
+
+	res map[string]*metrics.EventMetrics  // Results by target
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
+func (p *Probe) Init(name string, opts *options.Options) error {
```
Comment (alicjakwie) [inline GCS/probe0/probe0.go]:
```suggestion
func (p *HermesProbe) Init(name string, opts *options.Options) error {
```

File: GCS/probe0/probe0.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	p.name = name
+	p.opts = opts
+	p.l = opts.Logger
+	p.res = make(map[string]*metrics.EventMetrics)
+
+	return nil
+}
+
+// Start starts and runs the probe indefinitely.
+func (p *Probe) Start(ctx context.Context, dataChan chan *metrics.EventMetrics) {
```
Comment (alicjakwie) [inline GCS/probe0/probe0.go]:
```suggestion
func (p *HermesProbe) Start(ctx context.Context, dataChan chan *metrics.EventMetrics) {
```

File: GCS/probe0/probe0.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+			p.initProbeMetrics()
+			probeCtx, cancelFunc := context.WithDeadline(ctx, time.Now().Add(p.opts.Timeout))
+			p.runProbe(probeCtx)
+			cancelFunc()
+		}
+	}
+}
+
+// initProbeMetrics initializes missing probe metrics.
+func (p *Probe) initProbeMetrics() {
```
Comment (alicjakwie) [inline GCS/probe0/probe0.go]:
```suggestion
func (p *HermesProbe) initProbeMetrics() {
```

File: GCS/probe0/probe0.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+			AddMetric("total", metrics.NewInt(0)).
+			AddMetric("success", metrics.NewInt(0)).
+			AddMetric("latency", latVal).
+			AddLabel("probe", p.name).
+			AddLabel("dst", target)
+	}
+}
+
+// runProbeForTarget runs probe for a single target.
+func (p *Probe) runProbeForTarget(ctx context.Context, target string) error { //'CHANGE THIS
```
Comment (alicjakwie) [inline GCS/probe0/probe0.go]:
```suggestion
func (p *HermesProbe) runProbeForTarget(ctx context.Context, target string) error {
```

File: GCS/probe0/probe0.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+		 if err := wc.Close(); err != nil {
+				 return fmt.Errorf("Writer.Close: %v", err)
+		 }
+		 fmt.Fprintf(w, "Blob %v uploaded.\n", object)
+		 return nil
+
+}
+
+// runProbe runs probe for all targets and update EventMetrics.
+func (p *Probe) runProbe(ctx context.Context) {
```
Comment (alicjakwie) [inline GCS/probe0/probe0.go]:
```suggestion
func (p *HermesProbe) runProbe(ctx context.Context) {
```

File: GCS/probe0/probe0.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
+import (
```
Comment (alicjakwie) [inline GCS/probe0/probe0.go]:
```suggestion
package probe0

import (
```

File: GCS/probe0/probe0.go [hunk 13]
```diff
@@ -0,0 +1,143 @@
+package probe0
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Add package description:

// Package probe0 ...

File: GCS/probe0/probe0.go [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"io"
+	"os"
+	"time"
+	"cloud.google.com/go/storage"
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Nit: add an empty line above this to create two import groups:

1) Standard library packages
2) Other packages

File: GCS/probe0/probe0.go [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/google/cloudprober/targets/endpoint"
+	"github.com/googleinterns/step224-2020/blob/evan/hermesMain/cmd"
+)
+
+// Probe holds aggregate information about all probe runs, per-target.
+type HermesProbe struct {
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Why is this "HermesProbe"? Are there other possible probe types?
Is this type going to be used in all probes in Hermes?

File: GCS/probe0/probe0.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+	"time"
+	"cloud.google.com/go/storage"
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/google/cloudprober/targets/endpoint"
+	"github.com/googleinterns/step224-2020/blob/evan/hermesMain/cmd"
+)
+
+// Probe holds aggregate information about all probe runs, per-target.
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Comments should start with the exact name of the symbol they refer to: s/Probe/HermesProbe/.

File: GCS/probe0/probe0.go [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+type HermesProbe struct {
+	name    string
+	c       *ProbeConf
```
Comment (leczb) [inline GCS/probe0/probe0.go]: `c` is too terse here. How about `conf`?

File: GCS/probe0/probe0.go [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+	targets []string
+	opts    *options.Options
+
+	res map[string]*metrics.EventMetrics  // Results by target
```
Comment (leczb) [inline GCS/probe0/probe0.go]: `res` is too terse here. How about `results`?

File: GCS/probe0/probe0.go [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
+func (p *HermesProbe) Init(name string, opts *options.Options) error {
```
Comment (leczb) [inline GCS/probe0/probe0.go]: The convention is to define a stad-alone New() function that does this.

func New(name string, opts *options.Options) (*HermesProbe, error)

File: GCS/probe0/probe0.go [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+	p.c, ok := opts.ProbeConf.(*ProbeConf)
+	if !ok {
+		return fmt.Errorf("Invalid Hermes probe configuration")
```
Comment (leczb) [inline GCS/probe0/probe0.go]: I would include an explanation why it was deemed invalid:

fmt.Errorf("invalid argument: the type of opts.ProbeConf is not *ProbeConf") ...or something to that effect.

Also note that error messages should start with lower-case.

File: GCS/probe0/probe0.go [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+	p.name = name
+	p.opts = opts
+	p.l = opts.Logger
+	p.res = make(map[string]*metrics.EventMetrics)
+
+	return nil
+}
+
+// Start starts and runs the probe indefinitely.
+func (p *HermesProbe) Start(ctx context.Context, dataChan chan *metrics.EventMetrics) {
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Is `dataChan` a bidirectional channel? If not, indicate its direction.

File: GCS/probe0/probe0.go [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+	}
+	p.name = name
+	p.opts = opts
+	p.l = opts.Logger
+	p.res = make(map[string]*metrics.EventMetrics)
+
+	return nil
+}
+
+// Start starts and runs the probe indefinitely.
```
Comment (leczb) [inline GCS/probe0/probe0.go]: I would comment that this is not thread-safe.

File: GCS/probe0/probe0.go [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+			AddLabel("dst", target)
+	}
+}
+
+// runProbeForTarget runs probe for a single target.
+func (p *HermesProbe) runProbeForTarget(ctx context.Context, target string) error {
+	bucket := p.c.GetBucket()
+ 	object := p.c.GetObject()
+	client, err := storage.NewClient(ctx)
+          if err != nil {
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Nit: fix the indentation.

File: GCS/probe0/probe0.go [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+                  return fmt.Errorf("storage.NewClient: %v", err)
+          }
+          defer client.Close()
+ 
+         ctx, cancel := context.WithTimeout(ctx, p.opts.Timeout)
+		 defer cancel()
+
+		 f, err := os.Open("notes.txt")
```
Comment (leczb) [inline GCS/probe0/probe0.go]: This seems out of place. Why open an external file?

File: GCS/probe0/probe0.go [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+ 
+		 // Upload an object with storage.Writer.
+		 wc := client.Bucket(bucket).Object(object).NewWriter(ctx)
+		 if _, err = io.Copy(wc, f); err != nil {
+				 return fmt.Errorf("io.Copy: %v", err)
+		 }
+		 if err := wc.Close(); err != nil {
+				 return fmt.Errorf("Writer.Close: %v", err)
+		 }
+		 fmt.Fprintf(w, "Blob %v uploaded.\n", object)
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Use the logger for logging.

File: GCS/probe0/probe0.go [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+			start := time.Now()
+			err := p.runProbeForTarget(ctx, target) // run probe just for a single target
+			if err != nil {
+				p.l.Errorf(err.Error())
+				return
+			}
+			em.Metric("success").AddInt64(1)
+			em.Metric("latency").AddFloat64(time.Now().Sub(start).Seconds() / p.opts.LatencyUnit.Seconds())
+		}(target, p.res[target])
+
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Nit: remove stray empty line.

File: GCS/probe0/probe0.go [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+			defer wg.Done()
+			em.Metric("total").AddInt64(1)
+			start := time.Now()
+			err := p.runProbeForTarget(ctx, target) // run probe just for a single target
+			if err != nil {
+				p.l.Errorf(err.Error())
+				return
+			}
+			em.Metric("success").AddInt64(1)
+			em.Metric("latency").AddFloat64(time.Now().Sub(start).Seconds() / p.opts.LatencyUnit.Seconds())
```
Comment (leczb) [inline GCS/probe0/probe0.go]: This means that the latency metric will only show the latency of successful attempts.
Is that the intention?
As a user I would expect the latency metric to have a "success" label, so I can slice and dice based on my needs. It's useful to have latency values for failed attempts as well, as we can distinguish between quick failures vs. slow ones.

File: GCS/probe0/probe0.go [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+				 return fmt.Errorf("os.Open: %v", err)
+		 }
+		 defer f.Close()
+ 
+		 // Upload an object with storage.Writer.
+		 wc := client.Bucket(bucket).Object(object).NewWriter(ctx)
+		 if _, err = io.Copy(wc, f); err != nil {
+				 return fmt.Errorf("io.Copy: %v", err)
+		 }
+		 if err := wc.Close(); err != nil {
```
Comment (leczb) [inline GCS/probe0/probe0.go]: What if io.Copy() or wc.Close() never returns?

File: GCS/probe0/probe0.go [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+	name    string
+	c       *ProbeConf
+	targets []string
+	opts    *options.Options
+
+	res map[string]*metrics.EventMetrics  // Results by target
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
```
Comment (leczb) [inline GCS/probe0/probe0.go]: Please explain what the arguments mean.

File: GCS/probe0/probe0.go (hunk 19 above)
Comment (evanSpendlove) [reply GCS/probe0/probe0.go]: Unfortunately, the `Init()` function is required as it is called within Cloudprober to initialise a new probe.

Source: https://github.com/google/cloudprober/blob/dd0bb3e2e8a54038da64c9b092db2f8941ecb69c/probes/probes.go#L189

File: GCS/probe0/probe0.go (hunk 4 above)
Comment (evanSpendlove) [reply GCS/probe0/probe0.go]: This will run as the proto message for `Probe0` is `ProbeConf`.

This line is casting an interface public field of a struct when accessing it (I think).

The field is defined here: https://github.com/google/cloudprober/blob/dd0bb3e2e8a54038da64c9b092db2f8941ecb69c/probes/options/options.go#L40

File: GCS/probe0/probe0.go (hunk 19 above)
Comment (evanSpendlove) [reply GCS/probe0/probe0.go:30]: This is because `Init()` is a function specified in the `Probe` interface that all must probes must implement.

File: GCS/nullprobe/nullprobe.go [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+	"sync"
+	"io"
+	"os"
+	"time"
+	"cloud.google.com/go/storage"
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/google/cloudprober/targets/endpoint"
+	"github.com/googleinterns/step224-2020/blob/evan/hermesMain/cmd"
```
Comment (deniadelcid) [inline GCS/nullprobe/nullprobe.go:15]: Double-check this import. It seems to include a path to a branch that the master branch shouldn't be referencing

File: GCS/nullprobe/nullprobe.go [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+// HermesProbe holds aggregate information about all probe runs, per-target
+type HermesProbe struct {
+	name    string
+	conf       *ProbeConf
+	targets []string
+	opts    *options.Options
+
```
Comment (deniadelcid) [inline GCS/nullprobe/nullprobe.go:24]: nit: I don't think you need this whitespace.
```suggestion
```

File: GCS/nullprobe/nullprobe.go [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+	results map[string]*metrics.EventMetrics  // Results by target
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
+func (p *HermesProbe) Init(name string, opts *options.Options) error {
+	p.conf, ok := opts.ProbeConf.(*ProbeConf)
+	if !ok {
+		return fmt.Errorf("Invalid argument: type of opts.ProbeConf does not match *ProbeConf")
```
Comment (deniadelcid) [inline GCS/nullprobe/nullprobe.go:33]: To make your testing simpler, you could make a little Enum of error types that you include in your errors. This way in your tests you could just compare error types as well as include the enum for your error in any relevant probe result/error/success metrics.

Totally optional though, you could also just open an issue to make this a future feature.

**Review by deniadelcid — COMMENTED** [review]
Can you run gofmt on this file?

File: GCS/nullprobe/nullprobe.go (hunk 32 above)
Comment (leczb) [reply GCS/nullprobe/nullprobe.go:33]: I would only recommend making the errors more elaborate if the caller has the handle different errors in different ways.

If that's the case, then use custom error types that implement the error interface (instead of enums).
See https://blog.golang.org/go1.13-errors.

**Review by cianbgoog — APPROVED** [review]
Just to check we still want this, given how the wider work is progressing?

**Comment by alicjakwie** [comment]
The contents of the probe0.go file are being integrated into probe.go and create_file.go hence this pull request is closed and  the comments from this PR will be addressed in the respective pull requests for probe.go and the CreateFile function
